File: Indexed Search/Hash Based Indexed Search/Hash Table Search/HashTableSearch.js
class HashTable {

    constructor() {

        this.SIZE = 10;

        this.table =
            Array
                .from(
                    {
                        length:
                        this.SIZE
                    },
                    () => []
                );
    }

    /*
        h(key)=key%SIZE
    */
    hash(key) {

        return (
            key %
            this.SIZE
        );
    }

    insert(key) {

        const index =
            this.hash(key);

        this.table[index]
            .push(key);
    }

    search(key) {

        return this.table[
            this.hash(key)
        ]
        .includes(key);
    }

    delete(key) {

        const index =
            this.hash(key);

        this.table[index] =
            this.table[index]
            .filter(
                x =>
                x !== key
            );
    }

    display() {

        this.table
            .forEach(
                (
                    bucket,
                    index
                ) => {

                    console.log(
                        `[${index}] `
                        +
                        bucket.join(
                            " -> "
                        )
                    );
                }
            );
    }
}

const ht =
    new HashTable();

ht.insert(10);
ht.insert(20);
ht.insert(35);

ht.display();

console.log(
    ht.search(20)
);

ht.delete(20);

ht.display();

File: Indexed Search/Hash Based Indexed Search/Hash Table Search/HashTableSearch.py
class HashTable:

    TABLE_SIZE = 10

    def __init__(self):

        self.table = [
            []
            for _ in range(
                self.TABLE_SIZE
            )
        ]

    # h(key)=key%TABLE_SIZE
    def hash(self, key):

        return (
            key %
            self.TABLE_SIZE
        )

    def insert(self, key):

        index = self.hash(key)

        self.table[index].append(key)

    def search(self, key):

        index = self.hash(key)

        return (
            key
            in
            self.table[index]
        )

    def delete(self, key):

        index = self.hash(key)

        if (
            key
            in
            self.table[index]
        ):
            self.table[index].remove(key)

    def display(self):

        for (
            index,
            bucket
        ) in enumerate(
            self.table
        ):

            print(
                f"[{index}] "
                +
                " -> ".join(
                    map(
                        str,
                        bucket
                    )
                )
            )


ht = HashTable()

ht.insert(10)
ht.insert(20)
ht.insert(35)

ht.display()

print(
    "\nSearch:",
    ht.search(20)
)

ht.delete(20)

ht.display()
